common/djangoapps/entitlements/models.py

```diff
@@ -24,3 +54,65 @@ class CourseEntitlement(TimeStampedModel):
         help_text='The current Course enrollment for this entitlement. If NULL the Learner has not enrolled.'
     )
     order_number = models.CharField(max_length=128, null=True)
+    _policy = models.ForeignKey(CourseEntitlementPolicy, null=True, blank=True)
+
+    @property
+    def expired_at_datetime(self):
+        if not self.expired_at and self.get_days_until_expiration() < 0:
+            self.expired_at = datetime.utcnow()
+            self.save()
+        return self.expired_at
+
+    @expired_at_datetime.setter
+    def expired_at_datetime(self, value):
+        self.expired_at = value
+
+    @property
+    def policy(self):
+        return self._policy or CourseEntitlementPolicy()
+
+    @policy.setter
+    def policy(self, value):
+        self._policy = value
+
+    def get_days_since_created(self):
+        """
+        Returns an integer of number of days since the entitlement has been created
+        """
+        utc = pytz.UTC
+        return (datetime.now(tz=utc) - self.created).days
+
+    def get_days_until_expiration(self):
+        """
+        Returns an integer of number of days until the entitlement expires
+        """
+        expiry_date = self.created + timedelta(self.policy.expiration_period_days)
+        now = datetime.now(tz=pytz.UTC)
+        return (expiry_date - now).days
+
+    def is_entitlement_regainable(self):
+        """
+        Determines from the policy if an entitlement can still be regained by the user, if they choose
+        to by leaving and regaining their entitlement within policy.regain_period_days days from start date of
+        the course or their redemption, whichever comes later
+        """
+        if self.enrollment_course_run:
+            now = datetime.now(tz=pytz.UTC)
+            course_overview = CourseOverview.get_from_id(self.enrollment_course_run.course_id)
+            return ((now - course_overview.start).days < self.policy.regain_period_days or
```

Comment:
Can `course_overview.start` be `None`? Maybe for a self-paced course? If so this will probably throw an error.

Reply:
While the column is nullable, in practice as far as I can tell it is a no. I checked on prod 
```
mysql> SELECT * FROM course_overviews_courseoverview WHERE start IS NULL;
Empty set (0.02 sec)
```

And this is out of 5722 rows.